src/extractor/guidance/suppress_mode_handler.cpp

```diff
@@ -0,0 +1,62 @@
+#include "extractor/guidance/constants.hpp"
+#include "extractor/guidance/suppress_mode_handler.hpp"
+#include "extractor/travel_mode.hpp"
+
+#include <algorithm>
+#include <boost/assert.hpp>
+
+namespace osrm
+{
+namespace extractor
+{
+namespace guidance
+{
+
+SuppressModeHandler::SuppressModeHandler(const IntersectionGenerator &intersection_generator,
+                                         const util::NodeBasedDynamicGraph &node_based_graph,
+                                         const std::vector<QueryNode> &node_info_list,
+                                         const util::NameTable &name_table,
+                                         const SuffixTable &street_name_suffix_table)
+    : IntersectionHandler(node_based_graph,
+                          node_info_list,
+                          name_table,
+                          street_name_suffix_table,
+                          intersection_generator)
+{
+}
+
+bool SuppressModeHandler::canProcess(const NodeID /*nid*/,
+                                     const EdgeID via_eid,
+                                     const Intersection &intersection) const
+{
+    // if the approach way is not on the suppression blacklist, there's no ways to suppress by this criteria
+    const auto in_mode = node_based_graph.GetEdgeData(via_eid).travel_mode;
+    const auto suppress_in_mode =
+        std::find(begin(SUPPRESS_MODE_LIST), end(SUPPRESS_MODE_LIST), in_mode);
+    return suppress_in_mode != end(SUPPRESS_MODE_LIST);
+}
+
+Intersection SuppressModeHandler::
+operator()(const NodeID /*nid*/, const EdgeID source_edge_id, Intersection intersection) const
+{
+    const auto in_mode = node_based_graph.GetEdgeData(intersection[0].eid).travel_mode;
+    BOOST_ASSERT(std::find(begin(SUPPRESS_MODE_LIST), end(SUPPRESS_MODE_LIST), in_mode) != SUPPRESS_MODE_LIST.end());
+
+    // if all out ways share the same mode as the approach way, suppress all the turns
+    const auto all_share_mode =
+        std::all_of(begin(intersection) + 1, end(intersection), [this, &in_mode](ConnectedRoad way) {
+            return node_based_graph.GetEdgeData(way.eid).travel_mode == in_mode;
+        });
+
+    if (all_share_mode)
```

Comment:
Originally I'd been selectively suppressing only instructions onto ways that involved modes on the mode blacklist, but this failed a test added by @MoKob wherein there is an intersection with a new ferry line and a land-based way.

https://github.com/Project-OSRM/osrm-backend/pull/3420/files#diff-0bf26f655163cdd60bedcd61e763028dR35

The test is written such that the instruction at the stopover point is expected to be suppressed. 

My thinking on this is that without a `destination` based instruction when the _first_ ferry line is boarded, it's not exactly useful to emit an instruction at the "stopover" if there really is a stopover at all or just a change of direction? Do we have a way of knowing like in the test above whether or not a switch to a new ferry involves user action? I actually would kind of assume that switching to a new ferry line necessitates user action, but I defer to @MoKob who wrote the original test.